src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -332,69 +157,125 @@ spinner(void *d)
 
 		rdtscll(now);
-		timer = tcap_cyc2time(now + 1000 * cyc_per_usec);
+		timer = tcap_cyc2time(now + GRANULARITY * cyc_per_usec);
 		cos_switch(tc, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, 0, timer, BOOT_CAPTBL_SELF_INITRCV_CPU_BASE,
-		           cos_sched_sync());
+			   cos_sched_sync());
 		p = c;
 		rdtscll(c);
-		if (i > 0) t += c - p;
+		if (i > 0) {
+			t += c - p;
+		}
 
-		while (cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_CPU_BASE, RCV_ALL_PENDING, 0,
-				     &rcvd, &tid, &blocked, &cycles, &thd_timeout) != 0)
-			;
+		clear_sched(&rcvd, &tid, &blocked, &cycles, &thd_timeout);
 	}
+	EXPECT_LLU_LT((long long unsigned)(t/T_ITER), (unsigned)(GRANULARITY * cyc_per_usec * MAX_THRS), "Test Timer MAX");
+	EXPECT_LLU_LT((unsigned)(GRANULARITY * cyc_per_usec * MIN_THRS), (long long unsigned)(t/T_ITER), "Test Timer MIN");
 
-	PRINTC("\tCycles per tick (1000 microseconds) = %lld, cycles threshold = %u\n", t / 16,
-	       (unsigned int)cos_hw_cycles_thresh(BOOT_CAPTBL_SELF_INITHW_BASE));
+	/* TIMER IN PAST */
+	thdid_t     tid;
+	int	    blocked, rcvd;
+	cycles_t    cycles, now;
+	tcap_time_t timer, thd_timeout;
+	c = 0, p = 0;
+
+	rdtscll(c);
+	timer = tcap_cyc2time(c - GRANULARITY * cyc_per_usec);
+	cos_switch(tc, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, 0, timer, BOOT_CAPTBL_SELF_INITRCV_CPU_BASE,
+		   cos_sched_sync());
+	p = c;
+	rdtscll(c);
+
+	EXPECT_LLU_LT((long long unsigned)(c-p), (unsigned)(GRANULARITY * cyc_per_usec), "Test Timer Past");
+
+	clear_sched(&rcvd, &tid, &blocked, &cycles, &thd_timeout);
+
+	/* TIMER NOW */
+	c = 0, p = 0;
+
+	rdtscll(c);
+	timer = tcap_cyc2time(c);
+	cos_switch(tc, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, 0, timer, BOOT_CAPTBL_SELF_INITRCV_CPU_BASE,
+		   cos_sched_sync());
+	p = c;
+	rdtscll(c);
+
+	EXPECT_LLU_LT((long long unsigned)(c-p), (unsigned)(GRANULARITY * cyc_per_usec), "Test Timer Now");
+
+	cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_CPU_BASE, RCV_ALL_PENDING, 0,
+		&rcvd, &tid, &blocked, &cycles, &thd_timeout)
+		;
+
+	EXPECT_LLU_LT((long long unsigned)cycles, (long long unsigned)(c-p), "Test sched_rcv");
+
+	clear_sched(&rcvd, &tid, &blocked, &cycles, &thd_timeout);
 
-	PRINTC("Timer test completed. Success.\n");
 }
 
 struct exec_cluster {
 	thdcap_t    tc;
 	arcvcap_t   rc;
-	tcap_t      tcc;
+	tcap_t	    tcc;
 	cycles_t    cyc;
-	asndcap_t   sc; /*send-cap to send to rc */
+	asndcap_t   sc;			/* send-cap to send to rc */
 	tcap_prio_t prio;
-	int         xseq; /* expected activation sequence number for this thread */
+	int	    xseq; 		/* expected activation sequence number for this thread */
 };
 
 struct budget_test_data {
-	/* p=parent, c=child, g=grand-child */
+					/* p=parent, c=child, g=grand-child */
 	struct exec_cluster p, c, g;
 } bt[NUM_CPU], mbt[NUM_CPU];
 
-static void
+static int
 exec_cluster_alloc(struct exec_cluster *e, cos_thd_fn_t fn, void *d, arcvcap_t parentc)
 {
 	e->tcc = cos_tcap_alloc(&booter_info);
-	assert(e->tcc);
+	if (EXPECT_LL_LT(1, e->tcc, "Cluster Allocation"))
```

Comment:
"Cluster Allocation" vs something that tells more about the exact step that failed, here TCap allocation. 
I'm not sure if we want "EXPECT" to replace all "asserts" in the code. In "test_thds", EXPECT on thd_alloc would make sense and not in "test_timer".

Reply:
The main reason, I replaced the `assert` was to not kill the tests if the `assert` happened in the main thd. It's just so execution can continue